File: setMisMatch.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findErrorNums = function(nums) {
    
    // cyclic sort
    let i=0;
    while(i<nums.length){
        let correct = nums[i] -1;
        if(nums[i] == nums[correct]) i++;
        else [nums[i],nums[correct]] = [nums[correct],nums[i]];
    }

    // find duplicate and missing
    for(let [index,element] of nums.entries()){
        if(element != index+1) return [element,index+1];
    }

    return [-1,-1];
};


console.log(findErrorNums([3,2,2]));

File: SetMisMatch.java
import java.util.Arrays;

public class SetMisMatch {
    public static void main(String[] args) {
        int[] nums = { 1, 2, 2, 4 };
        System.out.println(Arrays.toString(findErrorNums(nums)));
    }
    
    static public int[] findErrorNums(int[] nums) {
        int i=0;

        // cyclic sort
        while(i<nums.length){
            int correct = nums[i] -1;
            if(nums[i] == nums[correct]){
                i++;
            }
            else{
                int temp = nums[i];
                nums[i]  = nums[correct];
                nums[correct] = temp;
            }
        }

        // find duplicate and missing value
        for(int j=0; j<nums.length; j++){
            if(nums[j] != j+1){
                return new int[] {nums[j],j+1};
            }
        }
    
    return new int[] {-1,-1};
    }
}
